src/System.Security.Cryptography.Pkcs/src/System/Security/Cryptography/Pkcs/CmsSignature.cs

```diff
@@ -83,7 +92,21 @@ internal static CmsSignature Resolve(string signatureAlgorithmOid)
             }
 
             byte[] signature;
-            bool signed = processor.Sign(dataHash, hashAlgorithmName, certificate, silent, out oid, out signature);
+            bool signed = processor.Sign(dataHash, hashAlgorithmName, certificate, key, silent, out oid, out signature);
+
+            if (signed && key != null)
+            {
+                // signatureParameters is used to figure out padding for RSA and ignored for ECDsa and DSA
+                // For RSA we can only get Pkcs1 as Pss does not support signing and null will default to Pkcs1 on RSAPkcs1CmsSignature instance.
+                if (!processor.VerifySignature(dataHash, signature, oid.Value, hashAlgorithmName, signatureParameters: null, certificate))
```

Comment:
Would it be more efficient to just compare the public keys?

Reply:
> Would it be more efficient to just compare the public keys?

We'd need to write per-algorithm public key comparison operations, and we'd need to do all of the allocating for the arrays of the \[Alg\]Parameters structures.

Then there's a "the cert used EC-named-parameters, but the key for some reason only supports exporting as EC-explicit-parameters... do we reject the match?".

Signature verification seems simpler, and like it might be less overall work, on net.

Reply:
possibly but VerifySignature/Hash is a safe bet as there are no special cases - I can't tell that for sure when comparing bytes. @bartonjs had some opinion about it when I talked about it with him offline

Reply:
Ah, you already responded, comments are lagging here